File: offers/03/exp.go
package main

// 找出数组中的重复数字
// https://leetcode-cn.com/problems/shu-zu-zhong-zhong-fu-de-shu-zi-lcof/


import (
	"fmt"
	"sort"
)

// findRepeatNumber 借助Map
func findRepeatNumber(nums []int) int {
    repeatDict := make(map[int]bool)
    for i := range nums{
        if repeatDict[i] {
            return nums[i]
        } else {
            repeatDict[i] = true
        }
    }
    return 0
}

// findReatNumberV2 先排序后两两比较
func findReatNumberV2(nums []int)int{
	sort.Ints(nums)
	for i := range nums{
        if nums[i] == nums[i+1]  {
            return nums[i]
        }
    }
    return 0
}

// func findReatNumberV3(nums []int) int{
// 	for i := range nums{
// 		newSlice := append(nums[:i], nums[i+1:]...)
// 		fmt.Println(newSlice)
// 		fmt.Println(nums)
// 		idx := sort.SearchInts(newSlice, nums[i])
// 		fmt.Println(idx)
// 		if idx > len(newSlice) || newSlice[idx] != nums[i]{
// 			return nums[i]
// 		}    
// 	}
// 	return 0
// }

File: offers/03/exp.py
# -*- coding:utf-8 -*-
"""
找出数组中的重复数字
https://leetcode-cn.com/problems/shu-zu-zhong-zhong-fu-de-shu-zi-lcof/
"""


class Solution(object):
    def findRepeatNumberV1(self, nums):
        """
        借助Map空间换时间，空间复杂度O(n),时间复杂度O(n)
        :type nums: List[int]
        :rtype: int
        """
        repeat_dict = {}
        for i in nums:
            if i not in repeat_dict:
                repeat_dict[i] = 1
            else:
                return i

    def findRepeatNumberV2(self, nums):
        """
        先遍历，然后查询排除该元素的数组中是否还有该值
        """
        for i in xrange(len(nums)):
            n = nums[i]
            if n in (nums[:i] + nums[i + 1:]):
                return n

    def findRepeatNumberV3(self, nums):
        """
        先排序，两两比较
        """
        pass
